### PR TITLE
Module org/data.functor

### PR DESCRIPTION
**Pull request by Divesh-Otwani**
Depends on #275.

This PR moves Data.Functor.Internal.Traversable to it's own module and breaks up Data.Functor into internal modules.

- I thought that since `Traversable` depended on `Control` functors it made sense to make it its own module. I can move it back or we could move it to `Control.Functor` in the future if that makes more sense.
- I moved `void` from the `Prelude` to `Data.Functor.Linear` since it resolved a dependency and seemed to fit better there. Should I re-export it in the Prelude? 
- I made this depend on #275 since we need to be able to only have `Consumable` imported in the internal functor module to break an import loop.

### CODE REVIEW

File: linear-base.cabal
Comment (utdemir) [inline linear-base.cabal:41]: I don't think we should lose the word `Linear`. Although I am not sure if it should be `Data.Functor.Internal.Linear` or `Data.Functor.Linear.Internal`.

Comment (Divesh-Otwani) [reply linear-base.cabal:41]: This may seem a bit long, but I''ll change it to these:
- `Data.Functor.Linear.Internal.Functor`
- `Data.Functor.Linear.Internal.Applicative`

... since the most specific thing about these internal modules is the _concept_ they contain and the concept of `Applicative` is qualified, say, by being data, being a functor, being linear, and the module being internal.

In terms of interfaces, things like`Data.Functor.Linear` make sense as linear flavors of familiar concepts. But I think internal modules shouldn't follow the same pattern and instead have the last name be one central concept so that it's clear when we import them internally what chiefly we're importing.

File: src/Control/Monad/Linear.hs
Comment (aspiwack) [inline src/Control/Monad/Linear.hs:152]: It's completely unrelated, but do we have in our README, a documentation of our recommended qualifying conventions? That `Data.Functor.Linear` is typically qualified as `Data`?

File: src/Data/Functor/Internal/Linear/Applicative.hs
Comment (aspiwack) [inline src/Data/Functor/Internal/Linear/Applicative.hs:18]:
```suggestion
-- # Applicative definition
```

(mostly for the missing 'i', but I think that the capital 'D' is unwarranted as well)

File: src/Data/Traversable/Linear.hs
Comment (aspiwack) [inline src/Data/Traversable/Linear.hs:12]: I think I will disagree with this. `Traversable` is still a subclass of `Data.Functor`. A very natural one to boot. The fact that it mentions `Control.Functor` is completely irrelevant to the problem. The whole point of having internal modules is that we don't have to care about compilation order when we make interfaces. So I say: move `Traversable` back into `Data.Functor.Linear`.

File: src/Data/Functor/Linear.hs
Comment (aspiwack) [inline src/Data/Functor/Linear.hs:24]: It's a reasonable place for this. No need to re-export it in the prelude, since there can be a `Control.void` as well, and there is little reason to favour one over the other.

In an ideal world, the prelude would re-export the `Data` and `Control` namespace, but Haskell can't do that yet. So be it.

**Review by aspiwack — APPROVED** [review]
I want `Traversal` back into `Data.Functor.Linear` (see below). Apart from that, it does look pretty good. So no need for a second review round from me. Merge when ready.

File: src/Control/Monad/Linear.hs
Comment (Divesh-Otwani) [reply src/Control/Monad/Linear.hs:152]: No, I"ll add this in a later PR.

**Pull request merged**
